common/src/main/java/io/netty/util/AsciiString.java

```diff
@@ -36,10 +39,14 @@
  * {@link ByteBuffer}. It is often used in conjunction with {@link TextHeaders}.
  */
 public final class AsciiString extends ByteString implements CharSequence, Comparable<CharSequence> {
-
+    private static final HashCodeGenerator CASE_INSENSITIVE_HASHER =
+            PlatformDependent.hashCodeGeneratorAsciiCaseInsensitive();
     private static final char MAX_CHAR_VALUE = 255;
     public static final AsciiString EMPTY_STRING = new AsciiString("");
 
+    private String string;
+    private int caseInsensitiveHashCode = CASE_INSENSITIVE_HASHER.emptyHashValue();
```

Comment:
@nmittler - Your comment was collapsed but I think it is still worth discussion:

> Should this be volatile? Are you concerned about degradation of performance?

Maybe it needs to be volatile but the "desired" use case is to handle an "unmodifiable string" (this was the only use-case before `arrayOffset()` was added). In this case `caseInsensitiveHashCode` is initialized to some initial value, and then updated if the hashCode method is called...it may be computed multiple times, but should result in the same value.  I did add a `arrayChanged()` method for completeness if the array is directly accessed (via `array()`), or the underlying shared storage changes and we could get into trouble here I guess.  WDTY?